### PR TITLE
BI-4117 Model for strike off objections mongo db

### PR DESCRIPTION
**Pull request by mattch1**
Model for mongo database

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            this.eMail = eMail;
+            return this;
+        }
+
+        public Builder withRequestInformation(String requestInformation) {
+            this.requestInformation = requestInformation;
+            return this;
+        }
+
+        public Builder withCreatedOn(LocalDateTime createdOn) {
```
Comment (ICoxCH) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:41]: Could this field not be automatically filled in with LocalDateTime.now() when the build() method is called? or is there a requirement to set the createdOn field for anything other than now()?

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/RequestStatus.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package uk.gov.companieshouse.api.strikeoffobjections.model.request;
+
+public enum RequestStatus {
+    OPEN,
+    CHIPS_INCOMING_CONTACT_SENT,
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/RequestStatus.java:5]: we don't know how many contacts will be sent to chips yet so just have OPEN, SUBMITTED, PROCESSED for the time being.

File: pom.xml [hunk 3]
```diff
@@ -94,6 +94,22 @@
...
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-data-mongodb</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-jdbc</artifactId>
```
Comment (ICoxCH) [inline pom.xml:111]: I don't think we currently have any code that would require this dependency, we might need it later on but I would recommend removing until it's needed

File: src/test/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntityTest.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.function.Supplier;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class StrikeOffObjectionsRequestEntityTest {
+
+    private static final Supplier<LocalDateTime> MOCK_PRESENT = () -> LocalDateTime.of(2020, 6, 30, 10, 00);
+
+    @Test
+    public void test() {
```
Comment (ICoxCH) [inline src/test/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntityTest.java:15]: perhaps name this test something more descriptive, such as StrikeOffObjectionsRequestEntityBuilderTest

File: src/test/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntityTest.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.time.LocalDateTime;
+import java.util.function.Supplier;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class StrikeOffObjectionsRequestEntityTest {
+
+    private static final Supplier<LocalDateTime> MOCK_PRESENT = () -> LocalDateTime.of(2020, 6, 30, 10, 00);
```
Comment (ICoxCH) [inline src/test/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntityTest.java:12]: Is there a reason this is a Supplier and not just a LocalDateTime?

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
+package uk.gov.companieshouse.api.strikeoffobjections.model.request;
+
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.mapping.Document;
+import org.springframework.data.mongodb.core.mapping.Field;
+
+import java.time.LocalDateTime;
+
+@Document(collection = "strike_off_objections_requests")
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:9]: Mongo collection name is still to be defined (checking with Steve Roberts) but it's likely to not have "_requests" at the end. Will confirm once known.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            this.username = username;
+            return this;
+        }
+
+        public Builder withEMail(String eMail) {
+            this.eMail = eMail;
+            return this;
+        }
+
+        public Builder withRequestInformation(String requestInformation) {
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:36]: this should be withReason

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
+package uk.gov.companieshouse.api.strikeoffobjections.model.request;
+
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.mapping.Document;
+import org.springframework.data.mongodb.core.mapping.Field;
+
+import java.time.LocalDateTime;
+
+@Document(collection = "strike_off_objections_requests")
+public class StrikeOffObjectionsRequestEntity {
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:10]: So I think 'Request' will should probably go from here too.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+
+    public static class Builder {
+        private String requestId;
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:14]: And this is an 'objectionId' as per @markpit 's API spec PR.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        private String username;
+        private String eMail;
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:16]: I still think this field should be named eMailAddress.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            strikeOffObjectionsRequestEntity.setRequestInformation(this.requestInformation);
+            strikeOffObjectionsRequestEntity.setCreatedOn(this.createdOn);
+            strikeOffObjectionsRequestEntity.setStatus(this.status);
+            return strikeOffObjectionsRequestEntity;
+        }
+    }
+
+    @Id
+    private String id;
+    @Field("request_id")
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:65]: objection_id

Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:65]: we don't need both 'requestId' and 'id', so just use 'id'

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    private String requestId;
+    @Field("username")
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:67]: We'll need user info in a CreatedBy object
CreatedBy {
    String id;
    String forename;
    String surname;
    String email;
}

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    private String username;
+    @Field("e_mail")
+    private String eMail;
+    @Field("reason_information")
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:71]: just call "reason_information" and requestInformation "reason"

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java (hunk 11 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:65]: Do we know why we have them both in PTF?

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:65]: No seems strange, I can't see any reason for both being there

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java (hunk 6 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:9]: Ok. Confirmed that both Mongo DB name and collection name will be "strike_off_objections" so "_requests" should go.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java (hunk 9 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java:14]: I think it's just 'id' in the API spec

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java (hunk 1 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java]: We should delegate the responsibility for this to the service layer.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java (hunk 6 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java]: We'll keep what we have as a placeholder for now and chnage it when we have more of a decision.

Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsRequestEntity.java]: Yep. Decision made and approved by Steve Roberts - it's "strike_off_objections".

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
+package uk.gov.companieshouse.api.strikeoffobjections.model.request;
+
+public class CreatedBy {
+    private String id;
+    private String forename;
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java:5]: API has been updated. forename, surname were nice to have's and have been taken out. If we need names we can cross reference the user collection

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/RequestStatus.java [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package uk.gov.companieshouse.api.strikeoffobjections.model.request;
+
+public enum RequestStatus {
+    OPEN,
+    PROCESSED,
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/RequestStatus.java:5]: PROCESSED isn't in the spec (at least for now) so I think it should be removed from here.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java (hunk 14 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java]: Names are gone.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package uk.gov.companieshouse.api.strikeoffobjections.model.request;
+
+public class CreatedBy {
+    private String id;
+    private String emailAddress;
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java:5]: just 'email'

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java:5]: That's exactly what it was before.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+            strikeOffObjectionsEntity.setCreatedBy(this.createdBy);
+            strikeOffObjectionsEntity.setCreatedOn(this.createdOn);
+            strikeOffObjectionsEntity.setStatus(this.status);
+            return strikeOffObjectionsEntity;
+        }
+    }
+
+    @Id
+    private String id;
+    @Field("e_mail_address")
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java:58]: email is in createdBy

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java (hunk 16 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java:5]: good

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java (hunk 17 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java]: Gome form entity - changed to email in createdBy

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java]: Field annotations added to createdBy

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    @Id
+    private String id;
+    @Field("reason")
+    private String reason;
+    @Field("created_by")
+    private CreatedBy createdBy;
+    @Field("created_on")
+    private LocalDateTime createdOn;
+    @Field("status")
+    private RequestStatus status;
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java:58]: I think this entity is missing 'companyNumber'.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java (hunk 16 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java]: Not good.

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/CreatedBy.java]: it's what is used in other CH collections like accounts.users. We should get an official CH standard for email

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java (hunk 18 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java:67]: company number added.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/RequestStatus.java [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package uk.gov.companieshouse.api.strikeoffobjections.model.request;
+
+public enum RequestStatus {
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/RequestStatus.java:3]: This should probably be ObjectionStatus.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+        private CreatedBy createdBy;
+        private LocalDateTime createdOn;
+        private RequestStatus status;
+
+        public Builder withCompanyNumber(String companyNumber) {
+            this.companyNumber = companyNumber;
+            return this;
+        }
+
+        public Builder withReason(String requestInformation) {
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java:25]: Parameter name should be 'reason'.

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/StrikeOffObjectionsEntity.java]: renamed

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/RequestStatus.java (hunk 19 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/model/request/RequestStatus.java]: Renamed

**Pull request merged**